db/flush_job.cc

```diff
@@ -306,6 +317,272 @@ void FlushJob::Cancel() {
+            tombstone.end_key_,    // Value is end key.
+            nullptr,               // KV protection info set as nullptr since it
+                                   // should only be useful for the first add to
+                                   // the original memtable.
+            false,                 // : allow concurrent_memtable_writes_
+                                   // Not seen as necessary for now.
+            nullptr,               // get_post_process_info(m) must be nullptr
+                      // when concurrent_memtable_writes is switched off.
+            nullptr);  // hint, only used when concurrent_memtable_writes_
+                       // is switched on.
+
+        if (!s.ok()) {
+          break;
+        }
+
+        // If new_mem is full, add it to purged_mems, and allocate new
+        // memtable.
+        if (new_mem->ShouldScheduleFlush()) {
+          // Rectify the first sequence number, which (unlike the earliest seq
+          // number) needs to be present in the new memtable.
+          new_mem->SetFirstSequenceNumber(new_first_seqno);
+          purged_mems.push_back(new_mem);
+          new_mem = new MemTable(
+              (cfd_->internal_comparator()), *(cfd_->ioptions()),
+              mutable_cf_options_, cfd_->write_buffer_mgr(),
+              mems_[0]->GetEarliestSequenceNumber(), cfd_->GetID());
+          new_first_seqno = kMaxSequenceNumber;
+        }
+      }
+    }
+
+    // If everything happened smoothly and new_mem contains valid data,
+    // decide if it is flushed to storage or kept in the imm()
+    // memtable list (memory).
+    if (s.ok() && (new_first_seqno != kMaxSequenceNumber)) {
+      // Rectify the first sequence number, which (unlike the earliest seq
+      // number) needs to be present in the new memtable.
+      new_mem->SetFirstSequenceNumber(new_first_seqno);
+
+      // The new_mem is added to the list of immutable memtables
+      // only if it filled at less than 10% capacity (arbitrary heuristic).
+      if (new_mem->ApproximateMemoryUsage() <
+          static_cast<size_t>(
+              ceil(0.1 * mutable_cf_options_.write_buffer_size))) {
+        db_mutex_->Lock();
+        cfd_->imm()->Add(new_mem,
+                         &job_context_->memtables_to_free,
+                         false /* trigger_flush. Adding this memtable will not trigger any flush */);
+        new_mem->Ref();
+        db_mutex_->Unlock();
+      } else {
+        purged_mems.push_back(new_mem);
+      }
+    } else {
+      // In this case, the newly allocated new_mem is empty.
+      assert(new_mem != nullptr);
+      delete new_mem;
+    }
+  }
+
+  // Reacquire the mutex for WriteLevel0 function.
+  db_mutex_->Lock();
+
+  // If mempurge successful, don't write input tables to level0,
+  // but write any full output table to level0.
+  if (s.ok()) {
+    TEST_SYNC_POINT("DBImpl::FlushJob:MemPurgeSuccessful");
+    for (MemTable* m : mems_) {
```

Comment:
How are we sure that mems_ here is the same as above? Does the flush thread effectively take ownership of mems_? Can we make that an assertion, or simply not rely on that?

Reply:
Short answer: we are guaranteed that mems_ here is the same as above - no other flush thread can edit these memtables at the same time, and even additional operations like the `DB::Close` wait for all the flushes to happen before interfering with any of this. 
Long answer: `mems_` is a variable that belongs to the `FlushJob` object, and the `FlushJob` object is created inside each flush thread (details at db_impl_compaction_flush.cc:154). `mems_` simply contains the pointer addresses of the `imm()` memtables.
`mems_` is populated by the column family data object when the db_mutex is held (`FlushJob::PickMemTable`). Upon populating `mems_`, the `cfd` object mark that these memtables are being flushed by a flush thread somewhere ("`flush_in_progress_ = true`", memtable_list.cc:357), which guarantees that nothing happens to the memtables contained in the `mems_` object while the flush job is handling them.
Therefore we can be sure that there is no race condition thanks to the `db_mutex`, and since the memtables are marked with "`flush_in_progress=true`", we know that these tables wont be edited while we are working on them in the flush process.